packages/frontend/src/components/Dialog/Dialog.tsx

```diff
@@ -29,13 +27,52 @@ const Dialog = React.memo<Props>(
     height,
     ...props
   }) => {
+    canEscapeKeyClose
+
+    const dialog = useRef<HTMLDialogElement>(null)
+
+    const onClick = canOutsideClickClose
+      ? (ev: React.MouseEvent<HTMLDialogElement>) => {
+          if (!dialog.current) {
+            return
+          }
+          ev.stopPropagation()
+          // that is the way to check if we clicked on dialog::backdrop
+          const rect = dialog.current.getBoundingClientRect()
+          const isInDialog =
+            rect.top <= ev.clientY &&
+            ev.clientY <= rect.top + rect.height &&
+            rect.left <= ev.clientX &&
+            ev.clientX <= rect.left + rect.width
+          if (!isInDialog) {
+            dialog.current.close()
+          }
+        }
+      : () => {}
+
+    const onClose = () => {
+      dialog.current!.style.display = 'none'
+      props.onClose()
+    }
+
+    const onCancel = (ev: React.BaseSyntheticEvent) => {
+      if (!canEscapeKeyClose) {
+        ev.preventDefault()
+      }
+    }
```

Comment:
Isn't this also supposed to be invoked when the "Cancel" button is pressed?

Reply:
I didn't go over all the dialogs to change how buttons behave with `formmethod="dialog"` (and I suppose we'll have to wrap dialog content inside `<form>`), so cancel doesn't matter much here, the dialog closes anyway.

Also there's this fine line between close and cancel, where we probably have to discuss our policy, but after we clean up all the issues with dialog chaining. What I also didn't touch is the dialog return value.